### ISSUE TITLE
vault refresh broken with chef 12.4.0

### ISSUE THREAD

**Issue by dzabel** [chef/chef-vault#175]
just updated to chef 12.4.0 and now at least knife vault refresh is broken.
tried with 2.6.0 and gem build directly from latest git repo.

ERROR: TypeError: no implicit conversion of nil into String

```
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item_keys.rb:37:in `initialize': no implicit conversion of nil into String (TypeError)
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item_keys.rb:37:in `new'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item_keys.rb:37:in `add'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item.rb:477:in `add_client'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item.rb:86:in `block in clients'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item.rb:80:in `each'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item.rb:80:in `clients'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef-vault/item.rb:353:in `refresh'
	from /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-vault-2.6.1/lib/chef/knife/vault_refresh.rb:39:in `run'
	from /opt/chef/embedded/apps/chef/lib/chef/knife.rb:425:in `block in run_with_pretty_exceptions'
	from /opt/chef/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
	from /opt/chef/embedded/apps/chef/lib/chef/knife.rb:424:in `run_with_pretty_exceptions'
	from /opt/chef/embedded/apps/chef/lib/chef/knife.rb:215:in `run'
	from /opt/chef/embedded/apps/chef/lib/chef/application/knife.rb:142:in `run'
	from /opt/chef/embedded/apps/chef/bin/knife:25:in `<top (required)>'
	from /usr/bin/knife:53:in `load'
	from /usr/bin/knife:53:in `<main>'
```

same gem work with chef 12.3.0

**Comment by tylercloke** [comment]
This should fix you. https://github.com/chef/chef/pull/3630

Release coming shortly.

**Comment by jeremymv2** [comment]
How do I get the 12.4.1 fix for the Chef::User interface when I only have ChefDK installed (not chef-client or the chef gem)?

[~/Devel/ChefProject/chef/chef-repo]
jeremymiller-> which chef
/opt/chefdk/bin/chef
[~/Devel/ChefProject/chef/chef-repo]
jeremymiller-> chef -v
Chef Development Kit Version: 0.7.0.rc.0
chef-client version: 12.4.0
berks version: 3.2.4
kitchen version: 1.4.0
[~/Devel/ChefProject/chef/chef-repo]
jeremymiller->

**Comment by philoserf** [comment]
The latest chefdk rc has 12.4.1 in it.

**Comment by threezerous** [comment]
Was this fix merged in windows chef-client 12.4.1 and chefdk 0.7.0? I am still getting the same exact error though I have the chef-client 12.4.1 and chefdk 0.7.0 installed.

**Comment by jblaine** [comment]
This error persists in whatever chef-vault comes with ChefDK 0.9.0

**Comment by btm** [comment]
@tylercloke's fix for #176 went in Chef 12.5.1 / ChefDK 0.9.0. 

There's certainly more than one bug still.

**Comment by thommay** [comment]
@jblaine please provide more detail. 
- When did it break for you? 
- Does ChefDK 0.10.0 help at all. 
- Do you get a stack trace? Or nothing at all?

**Comment by jblaine** [comment]
* The date of my comment above is when we discovered the problem. I can't even remember the last time we ran refresh, so I have no data for you like "it broke between version X and Y"
* Don't know yet. Will try in the next week.
* Nothing.

**Comment by jblaine** [comment]
My single test just now with ChefDK 0.10.0 shows the refresh command working as expected.

1. Added new vault item with `-S 'chef_environment:development'`
1. Examined client list
1. Changed an existing random node to be in the `development` environment
1. Refreshed vault item
1. Confirmed client list was updated to include the new node

**Comment by thommay** [comment]
OK, I'm going to close this issue. Please create new issues if there are remaining problems with ChefDK 0.10.0

**Comment by threezerous** [comment]
I think the original issue was reported for chef 12.4.0 and AFAIK, the issue still exists in that version.....unless I missed some communication.

**Comment by thommay** [comment]
@threezerous right - and we've fixed the problem in a later (12.5.1) version of chef; 12.4.0 can't change because it's released.

**Comment by threezerous** [comment]
OK....thanks, will try with version 12.5.1
